### PR TITLE
[FLINK-18248][docs] Update data type documentation for 1.11

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

Updates the data types documentation according to current master.

## Brief change log

See commit messages.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? docs

### CODE REVIEW

File: docs/dev/table/types.md [hunk 1]
```diff
@@ -1125,9 +1131,110 @@ equivalent to `ROW<myField INT, myOtherField BOOLEAN>`.
...
+- All fields of the class must be readable by either `public` declaration or a getter that follows common
+coding style such as `getField()`, `isField()`, `field()`.
+- All fields of the class must be writable by either `public` declaration, fully assigning constructor,
+or a setter that follows common coding style such as `setField(...)`, `field(...)`.
+- All fields must be mapped to a data type either implicitly via reflective extraction or explicitly
+using the `@DataTypeHint` [annotations](#data-type-annotations).
+- Fields that are declared `static` or `transient` are ignored.
+
+The reflective extraction supports arbitrary nesting of fields as long as a field type does not
+(transitively) uses itself.
```
Comment (aljoscha) [inline docs/dev/table/types.md:1179]:
```suggestion
(transitively) refer to itself.
```

File: docs/dev/table/types.md [hunk 2]
```diff
@@ -1160,6 +1267,48 @@ DataTypes.BOOLEAN()
...
+</div>
+
+</div>
+
+The type can be declared using `RAW('class', 'snapshot')` where `class` is the originating class and
+`snapshot` is the serialized `TypeSerializerSnapshot` in Base64 encoding. Usually, the type string is not
+declared directly but is generated while persisting the type.
+
+In the API, the `RAW` type can be declared either by directly supplying a `Class` + `TypeSerializer` or
+by passing `Class` and let the framework extract `Class` + `TypeSerializer` from there.
```
Comment (aljoscha) [inline docs/dev/table/types.md:1302]:
```suggestion
by passing `Class` and letting the framework extract `Class` + `TypeSerializer` from there.
```

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/data/conversion/DataStructureConverters.java [hunk 3]
```diff
@@ -149,6 +147,10 @@
 		switch (logicalType.getTypeRoot()) {
 			case ARRAY:
 				return ArrayObjectArrayConverter.create(dataType);
+			case MULTISET:
```
Comment (aljoscha) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/data/conversion/DataStructureConverters.java:150]: Why were these moved?

Comment (twalthr) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/data/conversion/DataStructureConverters.java:150]: According to the definition of `MapType` and `MultisetType`. We also allow subclasses of `Map` such as `HashMap` or `LinkedHashMap` as input parameters.
```
putConverter(LogicalTypeRoot.MAP, Map.class, MapMapConverter::createForMapType);
```
... was too strict because it only matches if the class is `Map.class`.

Comment (aljoscha) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/data/conversion/DataStructureConverters.java:150]: Ok, understood!

**Comment by aljoscha** [comment]
This looks good to merge!
